[FEAT/#212] OTP 에러메세지 및 다이얼로그 추가와 소소한 개선

angryPodo opened the pull request:
## Related issue 🛠
- closed #212 

## Work Description ✏️
- 에러 메세지 추가
- 실패 다이얼로그 추가
- 상태 읽기 지연
- 파라미터 네이밍 변경

## Screenshot 📸

https://github.com/user-attachments/assets/9722c334-a714-4385-bbd2-57d3f4fb1769

## Uncompleted Tasks 😅
- [ ] 서버 붙이면 끝

## To Reviewers 📢
#209 의 작업을 가져왔기 때문에 Draft 걸었습니다. 앞선 PR머지 되면 리뷰해주세요.

angryPodo committed:
[ADD/#205] Create OTP feature module

- Add `presentation:otp` module to project.
- Include new module in `settings.gradle.kts`.
- Add `.gitignore` to `presentation/otp`.
- Create `AndroidManifest.xml` for `presentation:otp`.
- Add `espressoCore` dependency.

angryPodo committed:
[FEAT/#205] OtpTextField component

Implement OtpTextField composable for OTP input.
- Displays 6 input boxes for OTP digits.
- Handles input validation (digits only, max 6 characters).
- Changes appearance based on input state (empty, filled, error).
- Includes preview for different states.

angryPodo committed:
[mod/#205] Implement OtpTextField component

This commit introduces the `OtpTextField` composable, a custom UI element for OTP (One-Time Password) input.

- **`OtpTextField`**: The main composable function that handles OTP input and display.
    - Takes `otpText`, `onOtpTextChange`, and `isError` as parameters.
    - Uses `BasicTextField` for input and displays OTP characters in individual boxes.
- **`OtpChar`**: A private composable function responsible for rendering each character box within the `OtpTextField`.
    - Styles the box based on whether it's filled or in an error state.
    - Updates border color, background color, and text color accordingly.
- **Preview**: Includes a preview function (`OtpTextFieldPreview`) to demonstrate different states of the `OtpTextField` (empty, filled, error, full error).

angryPodo committed:
[MOD/#205] OtpTextField: Update otpText to lambda and handle keyboard done action

- Changed the `otpText` parameter in `OtpTextField` from a `String` to a `() -> String` lambda to align with Jetpack Compose best practices.
- Implemented `KeyboardActions` to clear focus when the "Done" action is triggered on the keyboard.
- Updated `KeyboardOptions` to specify `ImeAction.Done`.

angryPodo committed:
[FEAT/#205] Otp Screen and Navigation Setting

Commit:
[CI] Apply Spotless formatting

angryPodo committed:
[MOD/#205] Lint format

angryPodo committed:
Merge remote-tracking branch 'origin/feat/#205-otp-textfield' into feat/#205-otp-textfield

angryPodo committed:
Merge remote-tracking branch 'origin/feat/#205-otp-textfield' into feat/#208-otp-screen-ui

angryPodo committed:
[FEAT/#208] Add OtpScreen UI

This commit introduces the UI for the OTP (One-Time Password) screen.

Key changes:
- Added `OtpRoute` composable function to handle navigation.
- Implemented `OtpScreen` composable function which includes:
    - A `BackTopAppBar` for navigation.
    - An `OtpTextField` for OTP input.
    - Informational text about the OTP.
    - A clickable text for users who haven't received an OTP.
    - A "인증하기" (Authenticate) button.
- Added a preview function `OtpScreenPreview`.

angryPodo committed:
[MOD/#208] Modify OtpRoute navigation

angryPodo committed:
[DEL/#208] 에스프레소 의존성 삭제

angryPodo committed:
[MOD/#208] Make OtpScreen private and reorder parameters

The `OtpScreen` composable function is now private.
The `isButtonEnable` parameter has been moved to be the last function parameter.
The `Text` composable for "인증 번호가 오지 않았나요?" now directly applies `textDecoration`.

angryPodo committed:
[MOD/#208] Make OtpScreen private and reorder parameters

The `OtpScreen` composable function is now private.
The `isButtonEnable` parameter has been moved to be the last function parameter.
The `Text` composable for "인증 번호가 오지 않았나요?" now directly applies `textDecoration`.

angryPodo committed:
Merge remote-tracking branch 'origin/develop' into feat/#208-otp-screen-ui

# Conflicts:
#	presentation/main/src/main/java/com/hilingual/presentation/main/MainScreen.kt
#	presentation/otp/src/main/java/com/hilingual/presentation/otp/OtpScreen.kt
#	presentation/otp/src/main/java/com/hilingual/presentation/otp/navigation/OtpNavigation.kt

angryPodo committed:
[FEAT/#212] 인증 코드 에러 Text 컴포넌트 추가

angryPodo committed:
[FEAT/#212] 인증 5회 실패 다이얼로그 추가 및 코드개선

angryPodo committed:
[MOD/#212] Defer state read

angryPodo committed:
[MOD/#212] lint format

nahy-512 commented on presentation/otp/src/main/java/com/hilingual/presentation/otp/OtpScreen.kt:
```diff
@@ -46,16 +100,106 @@ fun OtpScreen(
+private fun OtpFailureDialog(
+    modifier: Modifier = Modifier,
+    onContactClick: () -> Unit,
+    onExitClick: () -> Unit
```
p1
onDismiss도 뚫려있어야 할 것 같아요!

angryPodo replied:
엥 저거 다이얼로그 사라지면 안될텐데?? 앱종료하기나 문의하기 두가지만 강제해야해요. Dismiss하면 또 입력할수있게하는건데?

nahy-512 replied:
아 맞네요 5회 이상 실패하면 더 시도할 수 없게 해야되는구나!! 확인했습니다
그럼 이건 단순 궁금증인데, dismiss 이벤트가 필요없는 경우에는 래핑하는 컴포넌트에 파라미터를 뚫어놓을 필요 없이 BasicDialog에 바로 dismiss를 넣어주는 게 좋을까요?

angryPodo replied:
이게 참 이해 못하겠는게...Dialog 기본 컴포에는 onDismiss가 필수란 말이죠?
근데 DialogProperties로 백드롭 자체를 막을수도있고....근데 nullable은 아니라서 람다는 넣어야하고
참 그렇습니다. 근데 이제 dismiss가 단순 다이얼로그를 닫는 용도가 아니라 닫힐때 해야하는 작업을 넣으라고 만들어놓은거긴 합니다만..

근데 BasicDialog에 바로 dismiss를 넣는다는 질문은 이해를 못했어요

nahy-512 replied:
아 지금 작성해주신 것처럼
```
TwoButtonDialog {
   onDismiss = { },
}
```
에서 onDismiss로 빈 람다 전달하는 부분을 여쭤본 거였습니다!! 제가 설명을 좀 헷갈리게 적었던 것 같네요ㅠㅠ

angryPodo replied:
그쵸 백버튼이랑 배경터치를 막으면 굳이 전달할 필요가 없긴합니다. 더 정확하게 하자면 다이얼로그가 사라지면서 할 작업이 없으면!

nahy-512 commented on presentation/otp/src/main/java/com/hilingual/presentation/otp/OtpScreen.kt:
```diff
@@ -46,16 +100,106 @@ fun OtpScreen(
+            dismissOnClickOutside = false
+        )
+    )
 }
```
p3
컴포넌트 패키지로 따로 분리하지 않으신 이유가 있을까요?

angryPodo replied:
Screen말고 재사용이 안되니까 투머치 분리라고 생각해서 패키지를 따로 안만들었어요. 이건 지금도 고민이긴합니다

nahy-512 commented on presentation/otp/src/main/java/com/hilingual/presentation/otp/OtpScreen.kt:
```diff
@@ -46,16 +100,106 @@ fun OtpScreen(
+            onContactClick = onContactClick,
+            onExitClick = onExitClick
+        )
+    }
```
p2
이거 안드 회의에서 얘기한 대로 다이얼로그에서 visible 처리해주기?

angryPodo replied:
좋습니다ㅎㅎ

nahy-512 commented on presentation/otp/src/main/java/com/hilingual/presentation/otp/OtpScreen.kt:
```diff
@@ -19,24 +19,78 @@ import androidx.compose.foundation.background
+    authFailureCountProvider: () -> Int,
+    onBackClicked: () -> Unit,
+    onNotReceivedCodeClick: () -> Unit,
+    onAuthButtonClick: () -> Unit,
```
p3
다른 데는 클릭이벤트 다 Button이 안 붙은 걸로 확인했는데, 특별히 얘만 붙인 이유가 있을까요?
하단에 꽉 차는 CTA 버튼이니까 다른 것보다 이벤트성이 강한 것 같긴 하다만.. 흠야

angryPodo replied:
네이밍때문에 보기가 어지러워가지고 특정하려고 넣긴했습니다만,,,,굳이같네요 onAuthClick으로 수정하겠습니다

angryPodo committed:
Merge remote-tracking branch 'origin/develop' into feat/#212-otptextfield-dialog-error-handling

# Conflicts:
#	presentation/otp/src/main/java/com/hilingual/presentation/otp/OtpScreen.kt

angryPodo committed:
[MOD/#212] Implement app exit on OTP failure

This commit implements the functionality to exit the app when OTP authentication fails 5 times.

- Modified `OtpScreen` to pass `context` to `OtpContent`.
- Updated `OtpContent` to call `(context as? Activity)?.finishAffinity()` when the exit button in the `OtpFailureDialog` is clicked.
- Refactored `OtpFailureDialog` to accept an `isVisible` parameter to control its visibility.
- Renamed `onAuthButtonClick` to `onAuthClick` in `OtpContent` for consistency.

nhyeonii commented on presentation/otp/src/main/java/com/hilingual/presentation/otp/OtpScreen.kt:
```diff
@@ -122,30 +163,48 @@ private fun OtpScreen(
     }
+
+    OtpFailureDialog(
+        isVisible = authFailureCountProvider() >= 5,
```
우와 어제 회의 때 이야기햇던 !! 이렇게 쓰는게 더 보기 좋은 것 같아요 👀👀

nhyeonii reviewed (approved):
고생하셨습니당 🚀 다이얼로그 저렇게 안에 뚫어서 관리하는게 뭔가 보기 더 깔끔한 느낌이네욤 !!

nahy-512 reviewed (approved):
고생하셨습니다🚀🚀

Daljyeong reviewed (approved):
확인 완료!! 회의 때 얘기했던 내용이 반영되어 있어서 참고할 때 많은 도움이 될 것 같아용
수고하셨습니닷 ☺️👍🏻👍🏻

Hyobeen-Park commented on presentation/otp/src/main/java/com/hilingual/presentation/otp/OtpScreen.kt:
```diff
@@ -122,30 +163,48 @@ private fun OtpScreen(
+    if (isVisible) {
+        TwoButtonDialog(
+            modifier = modifier,
+            title = "인증에 실패했어요",
```
진짜 사소한건데 . 빠짐

angryPodo replied:
매의 눈 미쳣다

Hyobeen-Park reviewed (approved):
어푸푸🚀🚀

angryPodo committed:
[MOD/#212] 라이팅 수정
